// @trezor/connect/lib/api/solana/additionalInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformAdditionalInfo = void 0;
const paramsValidator_1 = require("../common/paramsValidator");
const validateAdditionalInfo = (additionalInfo) => {
    var _a;
    (0, paramsValidator_1.validateParams)(additionalInfo, [{ name: 'tokenAccountsInfos', type: 'array' }]);
    (_a = additionalInfo.tokenAccountsInfos) === null || _a === void 0 ? void 0 : _a.forEach(tokenAccountInfo => {
        (0, paramsValidator_1.validateParams)(tokenAccountInfo, [
            { name: 'baseAddress', type: 'string', required: true },
            { name: 'tokenProgram', type: 'string', required: true },
            { name: 'tokenMint', type: 'string', required: true },
            { name: 'tokenAccount', type: 'string', required: true },
        ]);
    });
};
const transformAdditionalInfo = (additionalInfo) => {
    var _a;
    if (!additionalInfo) {
        return undefined;
    }
    validateAdditionalInfo(additionalInfo);
    return {
        token_accounts_infos: ((_a = additionalInfo.tokenAccountsInfos) === null || _a === void 0 ? void 0 : _a.map(tokenAccountInfo => ({
            base_address: tokenAccountInfo.baseAddress,
            token_program: tokenAccountInfo.tokenProgram,
            token_mint: tokenAccountInfo.tokenMint,
            token_account: tokenAccountInfo.tokenAccount,
        }))) || [],
    };
};
exports.transformAdditionalInfo = transformAdditionalInfo;
